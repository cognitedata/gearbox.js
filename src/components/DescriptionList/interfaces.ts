import { CSSProperties } from 'react';
import { AnyIfEmpty } from '../../interfaces';

export interface ValueListType {
  key?: string;
  name: string;
  value: string;
}

export interface MetaDescriptionListProps {
  /**
   * A custom categorization function (e.g. to collapse specific props)
   */
  toCategory?: (name: string) => string | undefined;
  /**
   * Categories to display on top. Can be used for sorting as well
   */
  categoryPriorityList?: string[];
  /**
   * A category name for uncategorised items
   */
  unknownCategoryName?: string;
  /**
   * Category names to be expanded by default
   */
  expandedCategories?: string[];
}

export interface DescriptionListProps extends MetaDescriptionListProps {
  /**
   * An object with properties to render
   */
  valueSet: { [name: string]: any };
  /**
   * An object with two properties
   * descriptionId and descriptionText to display above the table
   */
  description?: {
    descriptionId: string;
    descriptionText: string;
  };
  /**
   * Custom style object to style component wrapper
   */
  styles?: CSSProperties;
  /**
   * @ignore
   */
  theme?: AnyIfEmpty<{}>;
}
