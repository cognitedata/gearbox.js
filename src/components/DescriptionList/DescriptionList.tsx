import { Collapse } from 'antd';
import { Dictionary, groupBy } from 'lodash';
import React, { FC, Fragment } from 'react';
import styled from 'styled-components';
import { withDefaultTheme } from '../../hoc';
import { mapMetaData } from '../../utils/formatters';
import { DescriptionListProps, ValueListType } from './interfaces';
const { Panel } = Collapse;

const DescriptionList: FC<DescriptionListProps> = ({
  styles,
  valueSet,
  toCategory,
  description,
  expandedCategories,
  categoryPriorityList = [],
  unknownCategoryName = 'Uncategorised',
}: DescriptionListProps) => {
  const arrayValues = mapMetaData(valueSet);
  const categoriesMap: Dictionary<ValueListType[]> = groupBy<ValueListType>(
    arrayValues,
    toCategoryWithDefault
  );
  const categories = Object.keys(categoriesMap);

  function getPrioritizedCategoryName() {
    const rest = categories.filter(
      category => !categoryPriorityList.includes(category)
    );
    return [...categoryPriorityList, ...rest];
  }

  function getDefaultActiveKey() {
    return expandedCategories || [getPrioritizedCategoryName()[0]];
  }

  function toCategoryWithDefault({ name }: ValueListType) {
    let category;
    if (toCategory) {
      category = toCategory(name);
    }
    return category || unknownCategoryName;
  }

  const hasMultipleCategories = () => categories.length > 1;

  const renderProperty = (prop: ValueListType) => (
    <Fragment key={prop.key}>
      <dt>{prop.name}</dt>
      <dl>{prop.value}</dl>
    </Fragment>
  );

  const renderValues = (list: ValueListType[]) => {
    return (
      <DL
        style={styles}
        aria-describedby={description ? description.descriptionId : ''}
      >
        {list.map(renderProperty)}
      </DL>
    );
  };

  const renderList = () => {
    return hasMultipleCategories() ? (
      <Collapse defaultActiveKey={getDefaultActiveKey()}>
        {getPrioritizedCategoryName()
          .filter(categoryName => categoriesMap[categoryName])
          .map(categoryName => (
            <Panel key={categoryName} header={categoryName}>
              {renderValues(categoriesMap[categoryName])}
            </Panel>
          ))}
      </Collapse>
    ) : (
      renderValues(arrayValues)
    );
  };

  return (
    <>
      {description && (
        <p id={description.descriptionId}>{description.descriptionText}</p>
      )}
      {arrayValues.length ? renderList() : <NoData>No data</NoData>}
    </>
  );
};

const DL = styled('dl')`
  display: flex;
  flex-wrap: wrap;
  text-align: left;

  dl,
  dt {
    border: 1px solid #e8e8e8;
    border-bottom: unset;
    padding: 16px;
    width: 50%;
    margin: 0;
    background-color: white;
  }

  dl:last-of-type,
  dt:last-of-type {
    border-bottom: 1px solid #e8e8e8;
  }

  dt {
    border-right: unset;
    font-weight: 500;
  }

  @media only screen and (max-width: 840px) {
    dl,
    dt {
      width: 100%;
      border: 1px solid #e8e8e8;
      border-bottom: unset;
    }

    dt {
      background-color: #ececec;
    }

    dt:last-of-type {
      border-bottom: unset;
    }
  }
`;

const NoData = styled('p')`
  border: 1px solid #e8e8e8;
  padding: 16px;
  text-align: center;
`;

const Component = withDefaultTheme(DescriptionList);
Component.displayName = 'DescriptionList';

export { Component as DescriptionList };
