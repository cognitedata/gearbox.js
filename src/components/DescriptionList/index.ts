export * from './DescriptionList';
export * from './interfaces';
