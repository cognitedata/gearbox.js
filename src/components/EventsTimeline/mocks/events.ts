import { fakeEvents } from '../../../mocks';
import {
  CogniteEventForTimeline,
  EventTimelineType,
  EventTimelineView,
  TimelineEvent,
} from '../interfaces';

export const timelineEvents: TimelineEvent[] = [
  {
    id: 8825861064387,
    type: EventTimelineType.continuous,
    view: EventTimelineView.fill,
  },
  {
    id: 1995162693488,
    type: EventTimelineType.discrete,
    view: EventTimelineView.outline,
  },
];

export const getCogniteEventsForTimeline: (
  tlEvents: TimelineEvent[]
) => CogniteEventForTimeline[] = (tlEvents: TimelineEvent[]) =>
  tlEvents.map(({ id, type, view }) => ({
    ...fakeEvents.find(e => e.id === id)!,
    appearance: { view, type },
  }));
