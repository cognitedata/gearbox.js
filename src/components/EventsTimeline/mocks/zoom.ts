export const zoomMock = (zoomEvents: any = {}): any => {
  const self: any = {};

  self.scaleExtent = () => self;
  self.on = jest.fn().mockImplementation((name: string, fn: any) => {
    zoomEvents[name] = fn;

    return self;
  });
  self.apply = jest.fn();

  return self;
};
