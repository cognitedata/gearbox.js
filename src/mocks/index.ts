export * from './assets';
export * from './assetsList';
export * from './assetsListV2';
export * from './common';
export * from './datapoints';
export * from './documents';
export * from './events';
export * from './mockSdk';
export * from './threeD';
export * from './timeseriesListV2';
export * from './timeseriesDataPointsList';
export * from './threeDNodesList';
