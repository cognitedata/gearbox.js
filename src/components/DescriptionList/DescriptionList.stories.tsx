import React from 'react';
import { storiesOf } from '@storybook/react';
import DescriptionList from 'components/DescriptionList/DescriptionList';
import { ASSET_DATA } from 'mocks/assets';

storiesOf('DescriptionList', module)
  .add('Basic example', () => (
    <DescriptionList valueSet={ASSET_DATA.metadata} />
  ))
  .add('No data', () => <DescriptionList valueSet={[]} />)
  .add(
    'With description text',
    () => (
      <DescriptionList
        valueSet={ASSET_DATA.metadata}
        description={{
          descriptionId: 'list1',
          descriptionText: 'List with interesting data you might like to know.',
        }}
      />
    ),
    {
      info: {
        text:
          'The description-text above is linked to the list by aria-describedby. ' +
          'descriptionId will be the reference to the text - ' +
          'in case you have several description lists on one page.' +
          'You could also style it however you like with a specfic id-reference.',
      },
    }
  );
