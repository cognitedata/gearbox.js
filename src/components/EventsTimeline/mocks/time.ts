import { baseTimestamp } from '../../../mocks';

export const start = baseTimestamp;
export const end = baseTimestamp + 60 * 24 * 60 * 60 * 1000;
