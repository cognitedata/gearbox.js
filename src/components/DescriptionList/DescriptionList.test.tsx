import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';
import { ASSET_DATA } from '../../mocks';
import { DescriptionList } from './DescriptionList';

const ANT_COLLAPSE_HEADER = '.ant-collapse-header';
const ANT_COLLAPSE_CONTENT = '.ant-collapse-content';
const SAMPLE_PROPS = { 1: '1', 2: '1', 3: '2' };
const toCategoryByName = (name: string) =>
  Number.parseInt(name, 10) < 3 ? '1' : '2';

configure({ adapter: new Adapter() });

describe('AssetDetailsPanel', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <div>
          <DescriptionList valueSet={ASSET_DATA.metadata} />
          <DescriptionList valueSet={[]} />
        </div>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render description', () => {
    const tree = renderer
      .create(
        <div>
          <DescriptionList
            valueSet={ASSET_DATA.metadata}
            description={{
              descriptionId: '123',
              descriptionText: 'description',
            }}
          />
          <DescriptionList valueSet={[]} />
        </div>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render collapse for 2 categories', () => {
    const wrapper = mount(
      <DescriptionList valueSet={SAMPLE_PROPS} toCategory={toCategoryByName} />
    );
    wrapper
      .find(ANT_COLLAPSE_HEADER)
      .map(collapse => collapse.simulate('click'));
    expect(
      wrapper
        .find(ANT_COLLAPSE_CONTENT)
        .map(expand => expand.find('dt').map(item => item.text()))
    ).toEqual([['1', '2'], ['3']]);
  });

  it("shouldn't render collapse for 1 category", () => {
    const toCategory = () => 'category';
    const wrapper = mount(
      <DescriptionList valueSet={ASSET_DATA.metadata} toCategory={toCategory} />
    );
    expect(wrapper.exists(ANT_COLLAPSE_HEADER)).toBeFalsy();
  });

  it("shouldn't render collapse for 0 categories", () => {
    const wrapper = mount(<DescriptionList valueSet={ASSET_DATA.metadata} />);
    expect(wrapper.exists(ANT_COLLAPSE_HEADER)).toBeFalsy();
  });

  it('expands default categories', () => {
    const wrapper = mount(
      <DescriptionList
        valueSet={SAMPLE_PROPS}
        toCategory={toCategoryByName}
        expandedCategories={['2']}
      />
    );
    expect(
      wrapper
        .find(ANT_COLLAPSE_CONTENT)
        .find('dt')
        .text()
    ).toEqual('3');
  });

  it('renders uncategorised items', () => {
    const uncategorisedProp = 'latestUpdateTimeSource';
    const unknownCategory = 'extra';
    const all = 'all';
    const toCategory = (name: string) =>
      name !== uncategorisedProp ? all : undefined;
    const wrapper = mount(
      <DescriptionList
        valueSet={ASSET_DATA.metadata}
        toCategory={toCategory}
        unknownCategoryName={unknownCategory}
        expandedCategories={[unknownCategory, all]}
      />
    );
    expect(
      wrapper.find(ANT_COLLAPSE_HEADER).map(expand => expand.text())
    ).toEqual([all, unknownCategory]);
    expect(
      wrapper
        .find(ANT_COLLAPSE_CONTENT)
        .at(1)
        .find('dt')
        .text()
    ).toEqual(uncategorisedProp);
  });
});
