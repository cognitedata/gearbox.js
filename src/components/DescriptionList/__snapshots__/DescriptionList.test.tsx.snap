// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetDetailsPanel renders correctly 1`] = `
<div>
  <dl
    aria-describedby=""
    className="sc-htpNat UlplB"
  >
    <dt>
      SOURCE_DB
    </dt>
    <dl>
      workmate
    </dl>
    <dt>
      SOURCE_TABLE
    </dt>
    <dl>
      wmate_dba.wmt_location
    </dl>
    <dt>
      WMT_LOCATION_ID
    </dt>
    <dl>
      1004
    </dl>
    <dt>
      WMT_LOCATION_WORKSTART
    </dt>
    <dl>
      1999-09-01 07:00:00
    </dl>
    <dt>
      latestUpdateTimeSource
    </dt>
    <dl>
      1552471210000
    </dl>
  </dl>
  <p
    className="sc-bxivhb gmxaah"
  >
    No data
  </p>
</div>
`;

exports[`AssetDetailsPanel should render description 1`] = `
<div>
  <p
    id="123"
  >
    description
  </p>
  <dl
    aria-describedby="123"
    className="sc-htpNat UlplB"
  >
    <dt>
      SOURCE_DB
    </dt>
    <dl>
      workmate
    </dl>
    <dt>
      SOURCE_TABLE
    </dt>
    <dl>
      wmate_dba.wmt_location
    </dl>
    <dt>
      WMT_LOCATION_ID
    </dt>
    <dl>
      1004
    </dl>
    <dt>
      WMT_LOCATION_WORKSTART
    </dt>
    <dl>
      1999-09-01 07:00:00
    </dl>
    <dt>
      latestUpdateTimeSource
    </dt>
    <dl>
      1552471210000
    </dl>
  </dl>
  <p
    className="sc-bxivhb gmxaah"
  >
    No data
  </p>
</div>
`;
